rfcs/0005-fork-selection.md

```diff
@@ -0,0 +1,166 @@
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    4 |         4 |      2  |                    6 |
+|    3 |         3 |      4  |                    7 |
+|    2 |         2 |      8  |                   10 |
+|    1 |         1 |      16 |                   17 |
+
+*Vote 5* is at time 9, and the resulting state is
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    5 |         9 |      2  |                   11 |
+|    2 |         2 |      8  |                   10 |
+|    1 |         1 |      16 |                   17 |
+
+*Vote 6* is at time 10
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    6 |        10 |       2 |                   12 |
+|    5 |         9 |       4 |                   13 |
+|    2 |         2 |       8 |                   10 |
+|    1 |         1 |      16 |                   17 |
+
+At time 10 the new votes caught up to the previous votes.  But *vote 2* expires at 10, so the when *vote 7* at time 11 is applied the votes including and above *vote 2* will be popped.
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    7 |        11 |       2 |                   13 |
+|    1 |         1 |      16 |                   17 |
+
+The lockout for vote 1 will not increase from 16 until the stack contains 5 votes.
+
+### Slashing and Rewards
+
+The purpose of the lockout is to force a voter to commit opportunity cost to a specific fork.  Voters that violate the lockouts and vote for a diverging fork within the lockout should be punished.  Slashing or simply freezing the voter from rewards for a long period of time can be used as punishment.
+
+Voters should be rewarded for selecting the fork that the rest of the network selected as often as possible.  This is well-aligned with generating a reward when the vote stack is full and the oldest vote needs to be dequeued.  Thus a reward should be generated for each successful dequeue.
+
+### Economic Finality
+
+Economic finality can be calculated as the total opportunity costs due to the vote lockout at a given fork.  As a votes get further and further, buried the economic finality increases because the cost of unrolling would be the total loss of all the reward from the lockouts at that fork.
+
+### Thresholds
+
+Each voter can independently set a threshold of network commitment to a fork before that voter commits to a fork.  For example, at vote stack index 7, the lockout is 256 time units.  A voter may withhold votes and let votes 0-7 expire unless the vote at index 7 has at greater than 50% commitment in the network.  This allows each voter to independently control how much risk to commit to a fork.  Committing to a fork faster would allow the voter to earn more rewards.
```

Comment:
Re: "Committing to a fork faster". "faster" doesn't seem like the right word. Maybe "earlier"?

Reply:
at a higher frequency?